---
title: "my-vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(Damsel)
```

##thoughts
sketch out a format
-introduce data
-process data
-optional correlation plots
-then methylation analysis
-plots
-then finding peaks, any plots
-add the genes
-do the goseq


##1. Introduction
This document gives an introduction to the R package Damsel, for use in DamID analysis; from BAM file input to gene ontology analysis.
Designed for use with DamID data, the Damsel methodology could be modified for use on any similar technology that seeks to identify enriched regions relative to a control sample. Utilising the power of edgeR for differential analysis and goseq for gene ontology bias correction, Damsel provides a unique end to end analysis for DamID.

##2. Processing the BAM files
As a DamID analysis tool, Damsel requires a GATC region file for analysis. These regions serve as a guide to extract counts from the BAM files.

#2.1 Introducing the GATC region file
The GATC region file for Drosophila melanogaster v6 is available to you via data.
```{r}
#data("regions_gatc_drosophila_dm6")
#not sure how to write it actually
```
It is a dataframe with the consecutive GATC regions across the genome - representing the region (or the length) between GATC sites.

If you have another species of DamID data or would prefer to make your own region file, you can use the following fn:

If you already have your own GATC region file, ensure that it has the same format as `regions_gatc_drosophila_dm6` with 6 columns: 
-Position: chromosome-start
-seqnames: chromosome name
-start: start of region
-end: end of region
-width: length of region (ensure that is is correct according to [plyranges::as_granges()])

The GATC regions are not evenly distributed across the genome, and vary greatly in size. 
This can be explored in more detail via boxplots and #that one function I was making looking for patterns in the distribution


#2.2 Extracting the counts within the GATC regions
Note: Damsel requires BAM files that have been mapped to the reference genome.
Provided the path to a folder of BAM files (and their .bai files) and the appropriate GATC region file, the function `process_bams()` will extract the counts for each region for each available BAM and add them as columns to a data frame. The columns will be named by the BAM file name - please rename them before running `process_bams` if they do not make sense.
-the function makes use of [exomeCopy::countBaminGranges()]
```{r}
#process_bams(path, region = regions_gatc_drosophila_dm6)
```
-If necessary, at this stage please rearrange the BAM file columns so they are ordered in the following way: Dam_1, Fusion_1, Dam_2, Fusion_2 etc
-Do not remove the .bam extension on the column names as this is used as a check in later functions to ensure only the BAM files are selected from the dataframe.
```{r}
#example on rearranging columns
#df[,c(1:6,8,7,blah,blah)]
```

-The DamID data captures the ~75bp region extending from each GATC site, so although regions are of differing widths, there is a null to minimal length bias present on the data, and does not require length correction.


#2.3 Correlation analysis of samples
At this stage, the similarities and differences between the samples can be analysed via correlation.
 `corr_heatmap` plots the correlation of all available BAM files Dam and Fusion, to visualise the similarity between files.
The default for all Damsel correlation analysis is the non-parametric "spearman's" correlation.
The correlation between Dam_1 and Fusion_1 can be expected to reach ~ 0.7, whereas the correlation between Dam_1 & Dam_3 or Fusion_1 & Fusion_2 would be expected to be closer to ~0.9
```{r}
#corr_heatmap(df)
```

Two specific samples can also be compared using `corr_scatter` which plots a scatterplot of the two samples, overlaid with the correlation results. `corr_scatter` utilises [ggpubr::ggscatter()]
```{r}
#corr_scatter(df, sample_1 = col_1, sample_2 = col_2)
```


##3. Differential methylation analysis
The goal with DamID analysis is to identify regions that are enriched in the fusion sample relative to the control. In Damsel, this step is referred to as differential methylation analysis, and makes use of [edgeR].
For ease of use, Damsel has four main edgeR based functions which compile different steps and functions from within edgeR. 

#3.1 Setting up edgeR analysis
`edgeR_set_up` sets up the edgeR analysis for differential methylation testing. Taking the dataframe of samples and regions as input, it conducts the following steps:
-i. it extracts the sample data
-ii. groups the samples (Dam or Fusion)
-iii. filters the samples (remove regions with very low counts, the filtering parameters may be adjusted)
-iv. normalises the data
-v. establishes the design matrix (this includes the sample group and pairing replicates together - Dam_1 & Fusion_1)
-vi. estimates the dispersion

```{r}
#edgeR_set_up(df)
```
The output from this step is a grouped dataframe containing all of the information from the steps.

#3.2 Examining the data - multidimensional scaling plot
It's important to visualise the differences between the samples. This can be done with `edgeR_plot_mds`
You would expect the Dam samples to cluster together, and for the Fusion samples to cluster together.
You would expect the majority of the variation to be within the 1st dimension (the x axis), and less variation in the 2nd dimension (y axis)
```{r}
#edgeR_plot_mds(dge)
```

#3.3 Identifying differentially methylated regions
After exploring the data visually, it's time to identify the enriched regions. `edgeR_results` compiles the edgeR functions for differential testing with one key modification - it outputs the results with the adjusted p values instead of the raw p values.

`edgeR_results` conducts the following key steps:
-i. fits a QLF model - quasi likelihood
-ii. tests the model 
-iii. conducts p value adjustment and summarises model results by setting regions as either (1,0,-1) (log fold change and p value thresholds can be adjusted)

Outputted is a data frame with the region position, logFC (log fold change), logCPM (log counts per million), F (F statistic used to identify significance), PValue, adjustedP (tbd), dm (result: -1,0,1)

```{r}
#edgeR_results(dge)
```


#3.4 Plotting the results
The results from the above differential methylation analysis can be visualised in an MA style plot with `edgeR_results_plot`
MA style scatter plot with average logCPM on x-axis, average logFC on y-axis, with dots coloured by significance
```{r}
#edgeR_results_plot(results)
```


##4. Identifying peaks (bridges)
As you could see from the plot of the differential methylation results, there are 10s of 1000s of enriched regions. To reduce the scale of this data to something that can be more biologically meaningul, enriched regions can be compiled into peaks.

# ?Plotting
idk
# Aggregating the regions
Damsel approaches peaks with a simple approach - aggregating regions of result 1 together into peaks. Once this has been done, peaks or regions of 1s that have gaps of less than 150 bp are compiled together, and a note is made of the gap. Peaks are ranked by their mean p value - with regions that were exlcuded from edgeR analysis for being too small given a p value of 1.

Given that we know that Dam can methylate up to 5kb away, and most significantly within 1 kb, then peaks less than 2kb are considered "tightly bound",  2 kb < peaks < 10kb "bound", and peaks > 10kb "unexpected".

```{r}
#aggregate_peaks(dm_results)
```

# ?Plotting
-idk


##5. Identifying genes associated with peaks
The peak information itself - while interesting, has no biological meaning. As the peaks represent a region that the Fusion protein interacted with on the DNA, likely as a transcription factor, we wish to identify the gene that is being affected. To do so, we need to associate the peaks with a potential "target" gene.

Note: any gene identified here is only a potential target that must be validated in laboratory procedures. There is no method available that is able to accurately predict the location and target genes of enhancers, so a key and potentially incorrect assumption in this part of the analysis is that all peaks represent binding to a local enhancer or promoter - that it is close or overlapping to the target gene.

It must also be noted that the Drosophila melanogaster genome and transcription factor interactions are different to that of mammals and using the same assumptions is problematic but it's the best we have. While mammalian genes are generally spread out with little overlap, there is a large amount of overlap between Drosophila genes, requiring some intuitive interpretation of which gene the peak is potentially targeting.

In the Damsel methodology, peaks are considered to associate with genes if they overlap the gene body or are within 2.5kb upstream of the genes transcription start site. If multiple genes are within these criteria, they are all listed, with the closest gene given the primary position.

#5.1 Extract genes from Ensembl using biomaRt
Obtaining information about the genes is done through accessing Ensembl via biomaRt. As this can be a slow process, it has been provided for Drosophila melanogaster dm6 and can be loaded via `data`.

Functions to undertake this are also available (`get_biomart_genes`), and follow these general steps:
-i. accesses biomaRt using the seqnames of the appropriate GATC region file as a guide
-ii. accesses biomaRt a second time to obtain only the Ensembl canonical sequence information for each gene
-iii. identifies the number of GATC regions that overlap with each gene

```{r}
#data(gene_file)
#get_biomart_genes(species, version, regions)
```


#5.2 Examine distribution of genes and GATC regions
This can be visualised with ...

#5.3 Annotating genes to peaks
yeah I don't actually have a good fn for that yet
#5.4 Interpreting results/Plotting?
also don't have a good fn for that yet

##6. Gene ontology
#6.1 Setting up goseq
#6.2 Interpreting results




